usaspending_api/bulk_download/v2/views.py

```diff
@@ -0,0 +1,311 @@
+
+
+def verify_requested_columns_available(sources, requested):
+    bad_cols = set(requested)
+    for source in sources:
+        bad_cols -= set(source.columns(requested))
+    if bad_cols:
+        raise InvalidParameterException('Unknown columns: {}'.format(bad_cols))
+
+
+class BulkDownloadListAgenciesViewSet(APIView):
+    def post(self, request):
+        """Return list of agencies if no POST data is provided.
+        Otherwise, returns sub_agencies/federal_accounts associated with the agency provided"""
+        response_data = {'agencies': [],
+                         'sub_agencies': [],
+                         'federal_accounts': []}
+
+        agency_id = None
+        post_data = request.data
+        if post_data:
+            if 'agency' not in post_data:
+                raise InvalidParameterException('agency parameter not provided')
+            agency_id = post_data['agency']
+
+        # Get all the top tier agencies
+        toptier_agencies = list(ToptierAgency.objects.all().values('name', 'toptier_agency_id', 'cgac_code'))
+
+        if not agency_id:
+            # Return all the agencies if no agency id provided
+            response_data['agencies'] = toptier_agencies
+        else:
+            # Get the top tier agency object based on the agency id provided
+            top_tier_agency = list(filter(lambda toptier: toptier['toptier_agency_id'] == agency_id, toptier_agencies))
+            if not top_tier_agency:
+                raise InvalidParameterException('Agency ID not found')
+            top_tier_agency = top_tier_agency[0]
+            # Get the sub agencies and federal accounts associated with that top tier agency
+            response_data['sub_agencies'] = Agency.objects.filter(toptier_agency_id=agency_id)\
+                .annotate(subtier_agency_name=F('subtier_agency__name'),subtier_agency_id=F('subtier_agency__subtier_agency_id'))\
+                .values('subtier_agency_name', 'subtier_agency_id')
+            response_data['federal_accounts'] = FederalAccount.objects.filter(agency_identifier=top_tier_agency['cgac_code']) \
+                .annotate(federal_account_name=F('account_title'),
+                          federal_account_id=F('id'))\
+                .values('federal_account_name', 'federal_account_id')
+        return Response(response_data)
+
+
+class BulkDownloadAwardsViewSet(BaseDownloadViewSet):
+    """Generate bulk download for awards"""
+
+    def process_filters(self, filters, award_level):
+        """Filter function for Bulk Download Award Generation"""
+
+        for required_param in ['award_types', 'agency', 'sub_agency', 'date_type', 'date_range']:
+            if required_param not in filters:
+                raise InvalidParameterException('{} filter not provided'.format(required_param))
+
+        and_queryset_filters = {}
+
+        # Adding award type filter
+        award_types = []
+        if not isinstance(filters['award_types'], list):
```

Comment:
I recommend putting this in a try-except block and removing the type check.

Reply:
Sounds good. Fixed the other situations where this applies.